README
Makepp, a build program which has a number of features that allow for reliable
builds and simpler build files, is a drop-in replacement for GNU make. It
supports almost all of the syntax that GNU make supports, and can be used with
makefiles produced by utilities such as automake. It is called makepp (or
make++) because it was designed with special support for C++, which has since
been extended to other languages like Swig or embedded SQL. Also its
relationship to make is analogous to C++'s relationship to C: it is almost
100% backward compatible but adds a number of new features and much better
ways to write makefiles.

Makepp works on Unix/Linux-like systems, Ebcdic-based emulations like BS2000
or z/OS and Windows.  It is written entirely in Perl.  It requires version 5.6
of Perl or better.  It's distributed under the terms of the Perl Artistic
License or GPL 2 or newer, which means that it's freely available to anyone
for just about any purpose.

Some Perl versions have bugs that cause makepp to fail.  These are:
o Perl 5.6.0 fails in really weird ways on the x86 (and maybe other
  platforms) on large builds.
o Perl 5.6.1 has failed for us in different ways on different architectures.
  On some architectures it seems to work fine, on others it's flaky.

We develop and test mostly with Perl 5.8.0 or later, and we have
encountered no problems with it.

For the latest distribution, manual pages, and information about mailing
lists, please see the makepp home page, http://makepp.sourceforge.net.

Features:

o Makepp supports almost all makefile syntax that GNU make supports.
  Makepp is compatible enough with GNU make and stable enough to compile
  large packages, such as KDE 2.0.  (For details, see the manual section
  on old makefiles).

o Makepp automatically scans C/C++ sources for include files.  The
  implementation is flexible enough that support for other languages can
  be easily added.

o Makepp remembers the build command, and rebuilds if the command has
  changed, even if none of the files have changed.  This is useful if
  you change command line options (e.g., adding -DDEBUG or -g)--you
  don't have to do make clean.

o By default, makepp rebuilds if any dependency files have changed
  (even if they are still older than the target).  For example, if you
  replace a file with an older version, makepp knows to recompile.  This
  also means that makepp is not bothered by clock synchronization
  problems.  (If necessary, you can tell it to recompile only if the
  target is older than a dependency.)

o Makepp has a better system for handling makefiles spread across several
  directories.  makepp loads several makefiles simultaneously (keeping
  the variables and default directories separate), and it will realize
  automatically that a file needed by one makefile is built by a
  different makefile.  Recursive invocations of make are no longer
  necessary (but are still supported for backward compatibility).

o Makepp will not recompile if only comments or whitespace in
  C/C++ sources have changed.  This is especially important for header
  files which are automatically generated by other programs and are
  included in many modules.  Even if the date has changed because the
  file was remade, makepp won't recompile if the file hasn't changed.

o Makepp won't mix output from separate processes when running compilation
  commands in parallel.

o Makepp writes a log file explaining all of its build decisions,
  showing what each file depended on and what changed to cause each rule
  to be executed.  This makes debugging your makefiles much easier.

o Wildcards refer not only to files which already exist but to
  files which don't exist yet but which can be made.  Thus you can
  specify

    libmine.so: *.o

  and the compilation will work properly even if none of the
  .o files exist yet.

o Makepp can automatically separate your source and object directories,
  even if your makefile is not written that way.  (This is like VPATH
  but is completely transparent, even for include files.)  Makepp can
  also build separate copies of your program with different compilation
  options or on different architectures, automatically keeping the
  output files separate--you don't have to modify your makefiles at all.

o You can ask makepp to infer what object files are necessary for
  your program if your source files follow the simple convention that
  the implementation of all the classes and functions in a .h file is
  contained in a .cxx, .cc, .C, or .cpp file of the same name.  This can
  often avoid the use of .a files.

o Makepp won't be confused by multiple names for the same file.  It realizes
  that xyz.o, ./xyz.o, and ../objects/xyz.o are the same file, and if
  you load a makefile in a subdirectory, it will recognize that
  references in that subdirectory to ../xyz.o refer to the same file as
  xyz.o in the main build directory.  It also won't be confused by soft
  links in directory names.

o Makepp supports filenames with special characters like space or ":".  Just
  put any filenames with special characters in single or double quotes.

o Makepp supports more easily understood synonymns for the cryptic
  symbols $@, $<, $^, etc.

o You can define your own text processing functions in Perl if
  the default ones are not sophisticated or convenient enough.
  You can also embed Perl code to manipulate the make variables directly
  in your makefile.

Why not cons?  cook?  bake?  icmake?  jam?

These are good make utilities too, with some of the same features,
especially those relating to reliable builds.  I wrote makepp because
(at the time) cons wasn't flexible enough.

makepp is likely to work with existing makefiles, because it tries hard
to emulate all the features of the standard make implementations while
nonetheless being safer.  It has been tested successfully on several
large systems of makefiles, including large projects controlled by
autoconf and automake.  And you don't have to learn any new syntax or
write any new files to use makepp; you can probably use your old
makefiles (with possibly only minor modifications).  On the other hand,
you can take advantage of makepp's additional features to simply your
makefiles.


ChangeLog
2014-06-10  DP  <occitan@esperanto.org>

snapshot 2.0.98.5

	* Mpp/Cmds.pm: Eliminate new 5.20 warning, that CPAN counts as failed test.
	(c_template): New macro @include(filename)@.

	* Mpp/Makefile.pm (read_makefile)
	* Mpp/Subs.pm: Better show where errors come from.

2014-03-19  DP  <occitan@esperanto.org>

	* Mpp/FileOpt.pm (set_rule): A late found rule for a different dir can override a pattern rule.

2014-01-15  DP  <occitan@esperanto.org>

	* Mpp/CommandParser/Esql.pm
	* Mpp/CommandParser/Gcc.pm: Treat C_INCLUDE_PATH et al., -static and -l: and optimize.  Few API changes.

	* Mpp/Text.pm (split_path): Extra arg for parsing native vars even on Cygwin.

2013-12-01  DP  <occitan@esperanto.org>

	* Mpp/Text.pm:
	* Mpp/*.pm: Parens now nest in expressions.
	Single quotes now ignore \.
	unquote: no longer works on regexp vars.
	(r)find_unquoted: rename *index_ignoring_quotes.

2013-10-13  DP  <occitan@esperanto.org>

snapshot 2.0.98.4

	* config.pl
	* t/run_all.t
	* t/run_tests.pl
	* t/**/*test: Improve hinting and always show it, except in --test.  Try harder to get test results.

	* pod/html/html.pl (emit_item_tag): Put Compatibility with Incompatibilities.

2013-08-18  DP  <occitan@esperanto.org>

	* Mpp.pm
	* Mpp/FileOpt.pm
	* Mpp/Makefile.pm
	* Mpp/Repository.pm
	* Mpp/Rule.pm
	* Mpp/Scanner.pm
	* makepplog: New var MAKEPP_DEBUG and show effect of mppl -c

	* makepp_builtin_rules.mk: Evaluate some vars immediately.

2013-07-21  DP  <occitan@esperanto.org>

	* Mpp/Makefile.pm (grok_rule): Allow % to be inside make-expression.

	* Mpp/Text.pm (index_ignoring_single_quotes): Eliminate.
	(index_ignoring_quotes, split_on_whitespace): New type-option.

2013-07-05  DP  <occitan@esperanto.org>

	* install.pl
	* config.pl: Installation defaults for html-doc and man have changed slightly.

2013-05-19  DP  <occitan@esperanto.org>

	* pod/makepp_variables.pod: Document the details of rc-style
	substitution and how to deal with border cases.

	* pod/makepp_builtins.pod: Better structure and sort -t need not modify $_.

2013-05-09  DP  <occitan@esperanto.org>

	* Mpp.pm (perform): Continue --loop after error.

	* Mpp/Glob.pm: Don't fully qualify own $allow_dot_files.

	* Mpp/File.pm (mark_as_directory): Obtain FULLNAME only once.

	* makeppinfo: With --force show unremembered_SIGNATURE instead of warning.

2013-04-22  DP  <occitan@esperanto.org>

	* Mpp/FileOpt.pm (exists_or_can_be_built): Revert and improve, we do need to recurse.

	* t/run_tests.pl
	* t/**/*.test: Fix globbing for answers.  Check signatures after mpp.  New hook $mod_answer.

	* makepp (parse_command_line): Simpler way to avoid wrong "used once" warning.

	* config.pl: Win doesn't grok prefix export notation.

2013-04-14  DP  <occitan@esperanto.org>

	* makepp: Query build cache only if we use one.  Move some global bc and rep symbols and functions where they belong.

	* Mpp.pm (perform): Use map, to avoid spurious exit code.

	* Mpp/Rule.pm (DefaultRule::execute): Fix returning status.

	* Mpp/DB.pm: New debug module.

2013-03-30  DP  <occitan@esperanto.org>

	* makepp
	* Mpp/Rule.pm: find_all_targets_dependencies returns them already sorted.  sorted_dependencies wants list, not array.

	* Mpp/Subs.pm (infer_objects): optimize

	* makepp_builtin_rules.mk: Let infer_objects rules respect makepp_percent_subdirs.

	* install.pl
	* config.pl: Give more help about changing the perl binary.

2013-03-05  DP  <occitan@esperanto.org>

snapshot 2.0.98.3

2013-03-04  DP  <occitan@esperanto.org>

	* Mpp/Rule.pm (load_scaninfo_): Make private, don't force finding
	remembered deps, which may not be buildable any more
	(Mpp::DefaultRule::execute): No need for when_done.

	* Mpp/FileOpt.pm (exists_or_can_be_built): Handle repository files
	differently, since a dep there may not be valid here (hidden by
	chmod 0).
	(set_additional_dependencies): Remove and splice into grok_rule.

	* Mpp/Glob.pm: Optimize.

	* LICENSE: Add copyright because Debian wants it.

2013-02-16  DP  <occitan@esperanto.org>

	* *: Consistent and documented handling of perl instance including
	the possibility to install against /usr/bin/env.

	* Mpp/Rule.pm: Override-variable TMP.

2013-01-31  DP  <occitan@esperanto.org>

	* Mpp/Text.pm
	* makepp*: Fully move VERSION to Mpp::Text.

	* install.pl
	* config.pl: Fix help.

2013-01-20  DP  <occitan@esperanto.org>

	* Mpp/Recursive.pm
	* Mpp/Rule.pm
	* Mpp.pm: Move actual building to perform and implement --loop.

	* Mpp/Subs.pm (f_find_{first_}upwards): Don't find file above ROOT, if present.

	* additional_tests/2003_10_11_idash.test: Wait for timestamp only if something was built.

2013-01-08  DP  <occitan@esperanto.org>

	* install.pl: Gzip manfiles if local system does that, Debian policy.

2012-12-29  DP  <occitan@esperanto.org>

	* Mpp/Signature/c_compilation_md5.pm
	* t/md5.test: New option makepp_signature_C_flat.

2012-12-28  DP  <occitan@esperanto.org>

	* makepp
	* Mpp/Makefile.pm
	* Mpp/Recursive.pm
	* Mpp/Rule.pm
	* Mpp/Subs.pm
	* t/include.test: Defer decision about inexistant/stale include
	till end of makefile and reload if needed.

2012-11-12  DP  <occitan@esperanto.org>

	* Mpp/BuildCache.pm
	* makepp: Rename build_cache_error_hook to Mpp::BuildCache::error_hook.
	Move build_dependencies_done code to new Mpp::BuildCache::get.  Optimize.

	* Mpp.pm: Rename *hits to $Mpp::{BuildCache,Repository}::hits.
	Make final print readable.

	* Mpp/FileOpt.pm (grok_build_info_file): Rewrite as parser.

	* Mpp/Rule.pm (execute): Suppress stale rep symlink warning.

2012-11-02  DP  <occitan@esperanto.org>

	* *: All pre-2.0 features which issued deprecated-warnings are
	eliminated, as are $Mpp::Makefile::legacy_functions and
	$MAKEPP_INSTALL_OLD_MODULES.

2012-10-25  DP  <occitan@esperanto.org>

snapshot 2.0.98.2

	* Mpp/Text.pm: Handle new version scheme.

	* Mpp/Cmds.pm: Native Windows fails when closing --inpipe.


install.pl
#!/usr/bin/env perl
#
# This script asks the user the necessary questions for installing
# makepp and does some heavy HTML massageing.
#
# $Id: install.pl,v 1.120 2013/07/05 21:02:38 pfeiffer Exp $
#

package Mpp;

#
# First make sure this version of Perl is recent enough:
#
BEGIN { eval { require 5.008 } or exec $^X, 'config.pl' } # Dies with nice message.

BEGIN { eval { require Mpp::Text } or warn <<EOS and exit 1 } # avoid BEGIN/die diagnostic
Please call this script in the directory where you unpacked it!
EOS

use Config;
use File::Copy;
use File::Path;
use Mpp::File ();		# ensure HOME is set

system PERL, 'makepp', '--version'; # make sure it got a chance to apply workarounds.

print "\nUsing perl in " . PERL . ".\n";
print "If you want another, please set environment variable PERL to it & reinstall.\n"
  unless $ENV{PERL};

warn "\nMakepp is being installed with DOS newlines, as you unpacked it.\n\n"
  if ($_ = <DATA>) =~ tr/\r//;

our $eliminate = '';		# So you can say #@@eliminate

our $VERSION = $Mpp::Text::VERSION;
our $setVERSION = "  our \$VERSION = '$Mpp::Text::VERSION';";
our $BASEVERSION = $Mpp::Text::BASEVERSION;

#
# Now figure out where everything goes:
#
sub ARGV_or_prompt($$) {
  local $_ = shift @ARGV;
  return $_ if defined && 0 < length;

  local $| = 1;			# Enable autoflush on STDOUT.

  my $default = ref( $_[1] ) ? $_[1]() : $_[1];
  print "$_[0] [$default]? ";	# Print the prompt.
  $_ = <STDIN>;			# Read a line.
  s/^\s+//;
  s/\s+$//;
  return $default unless defined && 0 < length;
#
# Expand environment variables and home directories.
#
  my $orig = $_;
  s/\$(\w+)/$ENV{$1}/g;		# Expand environment variables.
  if (s/^~(\w*)//) {		# Is there a ~ expansion to do?
    if ($1 eq '') {
      $_ = "$ENV{HOME}$_";
    } else {
      my ($name, $passwd, $uid, $gid, $quota, $comment, $gcos, $dir, $shell) = getpwnam($1);
				# Expand from the passwd file.
      if ($dir) {		# Found it?
	$_ = "$dir$_";
      } else {
	$_ = "~$1";		# Not found.  Just put the ~ back.
      }
    }
  }
  print "  -> $_\n" if $orig ne $_;
  $_;
}

$prefix = '/usr/local';

$bindir = ARGV_or_prompt '
Makepp needs to know where you want to install it and its data files.
Makepp is written in Perl, but there is no particular reason to install
any part of it in the perl hierarchy; you can treat it as you would a
compiled binary which is completely independent of perl.

The questions understand environment variables like $HOME, or ~ syntax.

Where should the makepp executable be installed',
  "$prefix/bin";

$bindir =~ m@^(.*)/bin@ and $prefix = $1;
				# See if a prefix was specified.

my $datadir = ARGV_or_prompt '
Makepp has a number of library files that it needs to install somewhere.  Some
of these are Perl modules, but they can\'t be used by other Perl programs, so
there\'s no point in installing them in the perl modules hierarchy; they are
simply architecture-independent data that needs to be stored somewhere.

Where should the library files be installed',
  "$prefix/share/makepp";
our $setdatadir;
if ($datadir !~ /^\//) {	# Make a relative path absolute.
  use Cwd;
  my $cwd = cwd;
  chdir $datadir;
  $setdatadir = "\$datadir = '" . cwd . "';";
  chdir $cwd;
} else {
  $setdatadir = "\$datadir = '$datadir';";
}

$mandir = ARGV_or_prompt '
Where should the manual pages be installed?
Enter \"none\" if you do not want the manual pages.
Man directory',
  sub { my $d = "$prefix/share/man"; -d $d ? $d : "$prefix/man" };
our $noman = $mandir eq 'none' ? 1 : '';

$htmldir_val = $htmldir = ARGV_or_prompt '
Where should the HTML documentation be installed?
Enter "none" if you do not want any documentation installed.
HTML documentation directory',
  sub { my $d = "$prefix/share/doc"; -d $d ? "$d/makepp" : "$datadir/html" };

my $findbin = ARGV_or_prompt "
Where should the library files be sought relative to the executable?
Enter \"none\" to seek in $datadir",
  "none";
$findbin=0 if $findbin eq "none";
if($findbin) {
  $setdatadir = "use FindBin;\n" .
    qq{\$datadir = "\$FindBin::RealBin/$findbin";};
  $htmldir = qq{\$FindBin::RealBin/$findbin/html}
    if $htmldir eq $datadir . "/html";
}

my $destdir = shift @ARGV;

my %abbrev;			# which commands have a short form

if( $destdir ) {
  for( $bindir, $datadir, $mandir, $htmldir_val ) {
    s/^/$destdir/o if defined;
  }
}

mkpath "$datadir/$_" for
  qw(Mpp Mpp/ActionParser Mpp/BuildCheck Mpp/CommandParser Mpp/Fixer Mpp/Scanner Mpp/Signature);

$ENV{_MAKEPP_INSTALL} = 1;
substitute_file( $_, $bindir, 0755, 1 ) for
  qw(makepp makeppbuiltin makeppclean makeppgraph makeppinfo makepplog makeppreplay makepp_build_cache_control);

substitute_file( 'recursive_makepp', $datadir, 0755 );

substitute_file( 'Mpp/Text.pm', $datadir, 0644 );

foreach $module (qw(../Mpp

		    BuildCache BuildCacheControl Cmds Event File FileOpt Glob
		    Lexer Makefile Subs Repository Rule Utils

		    BuildCheck BuildCheck/architecture_independent
		    BuildCheck/exact_match BuildCheck/ignore_action
		    BuildCheck/only_action BuildCheck/target_newer

		    CommandParser CommandParser/Esql CommandParser/Gcc
		    CommandParser/Swig CommandParser/Vcs

		    Fixer/Automake Fixer/CMake

		    Scanner Scanner/C Scanner/Esqlc Scanner/Swig Scanner/Vera
		    Scanner/Verilog

		    Signature Signature/c_compilation_md5 Signature/md5
		    Signature/shared_object Signature/verilog_synthesis_md5
		    Signature/xml Signature/xml_space)) {
  copy("Mpp/$module.pm", "$datadir/Mpp/$module.pm");
  chmod 0644, "$datadir/Mpp/$module.pm";
}

our $explicit_perl = '';
{
  local $SIG{__WARN__}= sub {};
  $explicit_perl = "Mpp::PERL . ' ' ."
    if $destdir || system "$bindir/makeppinfo -qknone makepp"; # zero if executable, output nothing
}

substitute_file( 'Mpp/Recursive.pm', $datadir, 0644 );

foreach $include (qw(makepp_builtin_rules makepp_default_makefile)) {
  copy("$include.mk", "$datadir/$include.mk");
  chmod 0644, "$datadir/$include.mk";
}

#
# From here on we treat the pod files
#
chdir 'pod';

@pods = <*.pod>;

#
# Now massage and install the HTML pages.
#
if( $htmldir_val ne 'none' ) {
  {
    no warnings;
    @Mpp::html::pods = @pods;
    my $absre = is_windows ? qr/^\/|^[a-z]:/i : qr/^\//; # compensate chdir pod above
    $Mpp::html::target_dir = $htmldir_val =~ $absre ? $htmldir_val : "../$htmldir_val";
  }
  require 'html/html.pl';		# fix what Pod::Html does
  &Mpp::html::pods2html;
}

#
# Install the man pages:
#
if( $mandir ne 'none' ) {
  my $gzip = </usr/{,local/,gnu/}{,share/}man/**/*.[1-9]*.gz> ? '.gz' : ''; # Does this OS gzip them?
  mkpath "$mandir/man1";
  require Pod::Man;
  my %options = qw(errors pod
		   center Makepp
		   lax 1);
  my $parser = Pod::Man->new( %options );
  $abbrev{makepp_build_cache} = 'mppbcc'; # special case command documented with feature description
  for my $file (@pods) {
    next if $file eq 'makepp_index.pod'; # html only
    my $basename = substr $file, 0, -4;   # Get the name of the man file.
    my $manfile = "$mandir/man1/$basename.1";
    $parser->parse_from_file( $file, $manfile );
    chmod 0644, $manfile;
    system 'gzip', '-f9', $manfile and $gzip = '' if $gzip;
    my $abbrev = $abbrev{$basename};
    link "$mandir/man1/$basename.1$gzip", "$mandir/man1/$abbrev.1$gzip" or
      symlink "$basename.1$gzip", "$mandir/man1/$abbrev.1$gzip" if $abbrev;
  }
  link "$mandir/man1/makepp_build_cache.1$gzip", "$mandir/man1/makepp_build_cache_control.1$gzip" or
    symlink "makepp_build_cache.1$gzip", "$mandir/man1/makepp_build_cache_control.1$gzip";
}

print "makepp successfully installed.\n";

#
# This subroutine makes a copy of an input file, substituting all occurrences
# of @xyz@ with the Perl variable $xyz.  It also fixes up the header line
# "#!/usr/bin/env perl" if it sees one.  On Win also create a .bat to call it.
#
# Arguments:
# a) The input file.
# b) The output directory.
# c) The protection to give the file when it's installed.
# d) Create an mpp* abbreviation.
#
sub substitute_file {
  my ($infile, $outdir, $prot, $abbrev) = @_;

  local *INFILE;
  open INFILE, $infile or die "$0: can't read file $infile--$!\n";
  mkpath($outdir);

  open OUTFILE, "> $outdir/$infile" or die "$0: can't write to $outdir/$infile--$!\n";

  local $_;
  my $perl = PERL;
  while( <INFILE> ) {
    s@^#!\s*(/usr/bin/env )perl@$perl =~ tr!/!! ? "#!$perl -w" : "#!$1$perl"@oe
       if $. == 1;
    s/\\?\@(\w+)\@/$$1/g;		# Substitute anything containg @xyz@.
    if( /^#\@\@(\w+)/ ) {		# Substitute anything containg #@@xyz ... #@@
      1 until substr( <INFILE>, 0, 3 ) eq "#\@\@";
      $_ = $$1;
    }

    print OUTFILE $_;
    open INFILE, "$perl $infile --help |" # '-|' no good on Win AS
      if $abbrev && /__DATA__/;
  }
  close OUTFILE;
  close INFILE;

  chmod $prot, "$outdir/$infile";
  if( is_windows > 0 && $prot == 0755 ) {
    open my $outfile, "> $outdir/$infile.bat" or die "$0: can't write to $outdir/$infile.bat--$!\n";
    print $outfile "\@$perl $outdir/$infile %1 %2 %3 %4 %5 %6 %7 %8 %9\n";
    close $outfile;
  }
  if( $abbrev ) {
    $_ = $infile;
    {
      no warnings 'uninitialized';
      s/makepp(?:_?(.)[^_]+(?:_(.)[^_]+(?:_(.)[^_]+)?)?)?/mpp$1$2$3/;
    }
    $abbrev{$infile} = $_;
    link "$outdir/$infile", "$outdir/$_"
      or symlink $infile, "$outdir/$_";
    unless( -f "$outdir/$_" ) {
      copy "$outdir/$infile", "$outdir/$_";
      chmod $prot, "$outdir/$_";
    }
    copy "$outdir/$infile.bat", "$outdir/$_.bat" if is_windows > 0;
  }
}

__DATA__
Dummy line for newline test.
